{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "87cdecef-2c49-49f9-be9f-ce10f4842ef5",
      "metadata": {
        "id": "87cdecef-2c49-49f9-be9f-ce10f4842ef5"
      },
      "source": [
        "# Summary\n",
        "* So far, we have learned about all the basic syntax and tools available in Python. Specifically, we learned about:\n",
        "    * how break down a problem into 'sub problems'\n",
        "    * data types (strings, lists, tuples int etc) and their associated methods\n",
        "    * I/O (open with permission, read/write, close), input from user (input())\n",
        "    * loops, for and while\n",
        "    * Conditions: if, elif, else\n",
        "  \n",
        "* in Module 4B, we are going to increase the sophistication level of the tools that we incorporate into our programs.\n",
        "    * By choosing the best tool for our situation, instead of just using the fundamental tools of Python, we will improve the efficiency of our code."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a53c0c46-455d-42ed-8cfb-0a9ba0a53474",
      "metadata": {
        "id": "a53c0c46-455d-42ed-8cfb-0a9ba0a53474"
      },
      "source": [
        "## complex data types\n",
        "\n",
        "#### Warning: We will refine just ONE example for most of this notebook.\n",
        "__It is an example found in \"Python for Biologist\" by Martin Jones which is a clear text on introductory Python__\n",
        "\n",
        "* this is an attempt to demonstrate how much more efficient and clear the solution to the problem becomes as we use slightly more complex tools to solve it.\n",
        "\n",
        "* We often have paired data sets in biology\n",
        "    * Protein sequences and names\n",
        "    * DNA restriction enzymes and motifs\n",
        "    * Codons and their amino acid translation\n",
        "    \n",
        "* Dictionaries allow you to store and access paired data efficiently\n",
        "    * Dictionaries allow you to look up a word and a definition\n",
        "    * Of course, we could see what this type of problem looks like with a list…\n",
        "    \n",
        "* let's say we want to count trinucleotides present in the following sequence:\n",
        "\n",
        "        dna= \"ATCGATCGATCGTACGCTGA\"\n",
        "\n",
        "(BTW: creating a 'sliding glass window' that slides across a haplotype sequence - usually three nucleotides at a time, not just one - is extremely common in genomics and bioinformatics so this questions isn't as 'theoretical' or manufactored for the sake of an example as it might initially seem)\n",
        "\n",
        "-----------------------------\n",
        "\n",
        "What does the following program do?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0c328f35-c11f-4c4e-b49b-b244798d5503",
      "metadata": {
        "id": "0c328f35-c11f-4c4e-b49b-b244798d5503",
        "outputId": "92916819-e4f6-48ae-b205-977e436f1111"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "count is 0 for AAA\n",
            "count is 0 for AAC\n",
            "count is 0 for AAG\n",
            "count is 1 for AAT\n",
            "count is 0 for ACA\n",
            "count is 0 for ACC\n",
            "count is 1 for ACG\n",
            "count is 0 for ACT\n",
            "count is 0 for AGA\n",
            "count is 0 for AGC\n",
            "count is 0 for AGG\n",
            "count is 0 for AGT\n",
            "count is 0 for ATA\n",
            "count is 2 for ATC\n",
            "count is 1 for ATG\n",
            "count is 0 for ATT\n",
            "count is 0 for CAA\n",
            "count is 0 for CAC\n",
            "count is 0 for CAG\n",
            "count is 0 for CAT\n",
            "count is 0 for CCA\n",
            "count is 0 for CCC\n",
            "count is 0 for CCG\n",
            "count is 0 for CCT\n",
            "count is 1 for CGA\n",
            "count is 1 for CGC\n",
            "count is 0 for CGG\n",
            "count is 1 for CGT\n",
            "count is 0 for CTA\n",
            "count is 0 for CTC\n",
            "count is 1 for CTG\n",
            "count is 0 for CTT\n",
            "count is 0 for GAA\n",
            "count is 0 for GAC\n",
            "count is 0 for GAG\n",
            "count is 2 for GAT\n",
            "count is 0 for GCA\n",
            "count is 0 for GCC\n",
            "count is 0 for GCG\n",
            "count is 1 for GCT\n",
            "count is 0 for GGA\n",
            "count is 0 for GGC\n",
            "count is 0 for GGG\n",
            "count is 0 for GGT\n",
            "count is 1 for GTA\n",
            "count is 0 for GTC\n",
            "count is 0 for GTG\n",
            "count is 0 for GTT\n",
            "count is 0 for TAA\n",
            "count is 1 for TAC\n",
            "count is 0 for TAG\n",
            "count is 0 for TAT\n",
            "count is 0 for TCA\n",
            "count is 0 for TCC\n",
            "count is 2 for TCG\n",
            "count is 0 for TCT\n",
            "count is 2 for TGA\n",
            "count is 0 for TGC\n",
            "count is 0 for TGG\n",
            "count is 0 for TGT\n",
            "count is 0 for TTA\n",
            "count is 0 for TTC\n",
            "count is 0 for TTG\n",
            "count is 0 for TTT\n",
            "********************\n",
            "TTT\n",
            "--------------------\n",
            "[0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0]\n",
            "~~~~~~~~~~~~~~~~~~~~\n",
            "18\n",
            "CPU times: user 1.22 ms, sys: 766 µs, total: 1.99 ms\n",
            "Wall time: 1.58 ms\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "\n",
        "# %%time is a magic command that runs the entire cell multiple times and averages out how long it takes. It is not exactly the same\n",
        "# thing as %time which can have comments etc above it in the cell.\n",
        "# IT MUST BE THE VERY FIRST THING YOU TYPE IN A CELL - you can't even put comments above it --- I learned this the hard way.\n",
        "# There are MANY print statements in this code so the user time will be big and the system time - the time spent evaluating commands\n",
        "# by the kernel - will be small.\n",
        "\n",
        "# For loops revisited\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "# set up an empty list\n",
        "all_counts=[]\n",
        "# remember that we can loop through a list\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "# setting up the four bases as in line 16 and using THAT as the list to iterate over is best practice, but I am using lists here to illustrate.\n",
        "# This is because there is only ONE place in the code where you have to make changes. You'll see this in the next cell.\n",
        "# THIS ILLUSTRATES SCOPE NICELY, TOO since you can predict the order of trinucleotide creation. AAA, AAC, AAG, AAT, ACA, ACC, ACG, ACT...\n",
        "for base1 in [\"A\",\"C\",\"G\",\"T\"]:\n",
        "    for base2 in [\"A\",\"C\",\"G\",\"T\"]:\n",
        "        for base3 in [\"A\",\"C\",\"G\",\"T\"]:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            #print(trinucleotide)\n",
        "            count=dna.count(trinucleotide)\n",
        "            print(\"count is \"+str(count)+\" for \"+trinucleotide)\n",
        "            all_counts.append(count)\n",
        "\n",
        "print(\"*\"*20)\n",
        "# remember the idea of 'scope' of a variable? trinucleotide variable is only appropriate within the third loop. It will only print out\n",
        "# the last item it has in memory.\n",
        "print(trinucleotide)\n",
        "print(\"-\"*20)\n",
        "print(all_counts)\n",
        "print(\"~\"*20)\n",
        "print(sum(all_counts))"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "65d087c8-6550-4588-99d0-9b158ca8f90d",
      "metadata": {
        "id": "65d087c8-6550-4588-99d0-9b158ca8f90d"
      },
      "source": [
        "# Problem: we want a count of the trinucleotides present in the following sequence:\n",
        "    \n",
        "    dna=“ATCGATCGATCGTACGCTGA”\n",
        "\n",
        "* Using list/for loop:\n",
        "    * Output is inefficient (takes a long time) and hard to read – especially if you want to know about the frequency of a specific trinucleotide (such as ATG)\n",
        "    * We can try to do things like generate two lists (one of counts, one of trinucleotides) but this makes life more difficult because we have to keep track of TWO lists and make sure that the one-to-one correspondence of the elements remains the same. Let's modify the program above to do just that.\n",
        "\n",
        "# How might we solve this issue?"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f6906329-8a50-4422-9716-236b1c78f0cf",
      "metadata": {
        "id": "f6906329-8a50-4422-9716-236b1c78f0cf",
        "outputId": "730eda6f-ffc4-4c1c-d0d9-20d31005e632"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "[0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0]\n",
            "~~~~~~~~~~~~~~~~~~~~\n",
            "['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAA', 'TAC', 'TAG', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGA', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']\n",
            "CPU times: user 444 µs, sys: 162 µs, total: 606 µs\n",
            "Wall time: 566 µs\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "all_trinucleotides=[]\n",
        "all_counts=[]\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            #print(\"count is \"+str(count)+\" for \"+trinucleotide)\n",
        "            all_trinucleotides.append(trinucleotide)\n",
        "            all_counts.append(count)\n",
        "\n",
        "print(all_counts)\n",
        "print(\"~\"*20)\n",
        "print(all_trinucleotides)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "e60986ba-1327-414d-8877-814be20617fe",
      "metadata": {
        "id": "e60986ba-1327-414d-8877-814be20617fe"
      },
      "source": [
        "* our program has slightly more informative output but it still difficult to search for the count of the particular codon (trinucleotide) that you want to know\n",
        "* it is easy to imagine these two lists becoming separated or one list being modified but forgetting to modify the second list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e0138729-0d59-4e40-be7d-08707cb0f687",
      "metadata": {
        "id": "e0138729-0d59-4e40-be7d-08707cb0f687",
        "outputId": "9873db04-c78b-4abf-b3c3-4d802f24a1fb"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "count is 0 for AAA\n",
            "count is 0 for AAC\n",
            "count is 0 for AAG\n",
            "count is 1 for AAT\n",
            "count is 0 for ACA\n",
            "count is 0 for ACC\n",
            "count is 1 for ACG\n",
            "count is 0 for ACT\n",
            "count is 0 for AGA\n",
            "count is 0 for AGC\n",
            "count is 0 for AGG\n",
            "count is 0 for AGT\n",
            "count is 0 for ATA\n",
            "count is 2 for ATC\n",
            "count is 1 for ATG\n",
            "count is 0 for ATT\n",
            "count is 0 for CAA\n",
            "count is 0 for CAC\n",
            "count is 0 for CAG\n",
            "count is 0 for CAT\n",
            "count is 0 for CCA\n",
            "count is 0 for CCC\n",
            "count is 0 for CCG\n",
            "count is 0 for CCT\n",
            "count is 1 for CGA\n",
            "count is 1 for CGC\n",
            "count is 0 for CGG\n",
            "count is 1 for CGT\n",
            "count is 0 for CTA\n",
            "count is 0 for CTC\n",
            "count is 1 for CTG\n",
            "count is 0 for CTT\n",
            "count is 0 for GAA\n",
            "count is 0 for GAC\n",
            "count is 0 for GAG\n",
            "count is 2 for GAT\n",
            "count is 0 for GCA\n",
            "count is 0 for GCC\n",
            "count is 0 for GCG\n",
            "count is 1 for GCT\n",
            "count is 0 for GGA\n",
            "count is 0 for GGC\n",
            "count is 0 for GGG\n",
            "count is 0 for GGT\n",
            "count is 1 for GTA\n",
            "count is 0 for GTC\n",
            "count is 0 for GTG\n",
            "count is 0 for GTT\n",
            "count is 0 for TAA\n",
            "count is 1 for TAC\n",
            "count is 0 for TAG\n",
            "count is 0 for TAT\n",
            "count is 0 for TCA\n",
            "count is 0 for TCC\n",
            "count is 2 for TCG\n",
            "count is 0 for TCT\n",
            "count is 2 for TGA\n",
            "count is 0 for TGC\n",
            "count is 0 for TGG\n",
            "count is 0 for TGT\n",
            "count is 0 for TTA\n",
            "count is 0 for TTC\n",
            "count is 0 for TTG\n",
            "count is 0 for TTT\n",
            ".........................\n",
            "[0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 2, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 2, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 2, 0, 2, 0, 0, 0, 0, 0, 0, 0]\n",
            "~~~~~~~~~~~~~~~~~~~~~~~~~\n",
            "['AAA', 'AAC', 'AAG', 'AAT', 'ACA', 'ACC', 'ACG', 'ACT', 'AGA', 'AGC', 'AGG', 'AGT', 'ATA', 'ATC', 'ATG', 'ATT', 'CAA', 'CAC', 'CAG', 'CAT', 'CCA', 'CCC', 'CCG', 'CCT', 'CGA', 'CGC', 'CGG', 'CGT', 'CTA', 'CTC', 'CTG', 'CTT', 'GAA', 'GAC', 'GAG', 'GAT', 'GCA', 'GCC', 'GCG', 'GCT', 'GGA', 'GGC', 'GGG', 'GGT', 'GTA', 'GTC', 'GTG', 'GTT', 'TAA', 'TAC', 'TAG', 'TAT', 'TCA', 'TCC', 'TCG', 'TCT', 'TGA', 'TGC', 'TGG', 'TGT', 'TTA', 'TTC', 'TTG', 'TTT']\n",
            "*************************\n",
            "Here are the counts for TGA which is located at position: 56\n",
            "count for TGA is 2\n",
            "CPU times: user 881 µs, sys: 360 µs, total: 1.24 ms\n",
            "Wall time: 1.76 ms\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "#Let's add yet another level of sophistication to this example:\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "#trying to use lists and for loops - things that we have already learned about to- it is an inelegant solution\n",
        "all_trinucleotides=[]\n",
        "all_counts=[]\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            # you could print off the trinucleotide count as it is iterated here\n",
        "            # probably not the most efficient thing to do but still useful to\n",
        "            # understanding what the program is doing\n",
        "            print(\"count is \"+str(count)+\" for \"+trinucleotide)\n",
        "            all_trinucleotides.append(trinucleotide)\n",
        "            all_counts.append(count)\n",
        "\n",
        "print(\".\"*25)\n",
        "print(all_counts)\n",
        "print(\"~\"*25)\n",
        "print(all_trinucleotides)\n",
        "#we can look up the counts for specific trinucleotides\n",
        "i=all_trinucleotides.index(\"TGA\")\n",
        "c=all_counts[i]\n",
        "print(\"*\"*25)\n",
        "print(\"Here are the counts for TGA which is located at position: \"+str(i))\n",
        "print(\"count for TGA is \"+str(c))\n",
        "# of course, you could modify line 26 to include USER INPUT for whatever trinucleotide\n",
        "# you want specified for the count of  here."
      ]
    },
    {
      "cell_type": "markdown",
      "id": "a314ba74-2e80-4a55-8abb-08c77cc348a5",
      "metadata": {
        "id": "a314ba74-2e80-4a55-8abb-08c77cc348a5"
      },
      "source": [
        "There is still the problem of potential modification of one list but not the other since they aren't linked <-- **A REALLY BIG PROBLEM.** Python has a way to solve it, though!\n",
        "\n",
        "## Dictionaries\n",
        "* most important built-in Python data structure\n",
        "* Similar to a list but with __keys instead of an index__\n",
        "    * These are called key-value pairs\n",
        "    * more efficient use of memory than lists\n",
        "    * non-biological common use: email addresses (keys) and passwords (values)\n",
        "    \n",
        "* sometimes called \"associative arrays\" (less commonly: \"hash map\")\n",
        "*__SINCE PYTHON 3.6, DICTIONARIES ARE NOW 'INSERTION ORDERED' IN THE SAME WAY THAT LISTS ARE ORDERED. You should still know that traditionally, dictionaries are 'unordered' compared to lists since it is possible that you will use an older version of Python. Unordered means that when you looped through it, you could get a different order every time__\n",
        "\n",
        "* __Use {} instead of []__\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">\n",
        "Syntax:\n",
        "\t\t\td={‘key1’:1,’key2’:2,’key3’:3}\n",
        "\n",
        "* Key1 is key and points to 1 which is the value\n",
        "* Common (non-biological) uses include login user name/email address\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">\n",
        "Accessing/retrieving an element of a dictionary:\n",
        "\n",
        "\t\t\t\td['key2'] would result in 2\n",
        "\n",
        "## Creating a dictionary\n",
        "* You can add values to a dictionary (like you can with a list)\n",
        "* Adding a new key-value pair.\n",
        "    * Dictionaries come with the restriction that keys can only be certain data types: strings and numbers (so no file objects as keys) but values can be any type of data\n",
        "    * Keys must be **unique**\n",
        "    * Keys have to be **IMMUTABLE** - once set, they cannot be changed. We have only learned about two immutable types (strings and numbers) which limits our repetoire for keys at this point. We will learn about 'TUPLES' in the next lecture, though and they often act as dictionary keys.\n",
        "\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">\n",
        "Example:\n",
        "\n",
        "    #instantiate empty dictionary\n",
        "\n",
        "    menu ={}\n",
        "    \n",
        "    #add price to item\n",
        "\n",
        "    menu[\"Chicken Alfredo\"]=14.50\n",
        "    \n",
        "    len() can be used to determine the number of key-value pairs\n",
        "\n",
        "\n",
        "Example:\n",
        "\n",
        "    Zoo_animals={\"Unicorn\":\"Cotton Candy House\",\"Sloth\":\"Rainforest Exhibit\",\"Salmon\":\"Alaskan Estuary\"}\n",
        "\n",
        "‘Unicorn’ is the Key and ‘Cotton Candy House’ is the value (probably the location of the unicorn)\n",
        "\n",
        "* We can use for to loop through the keys of a dictionary\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">\n",
        "Example:\n",
        "\n",
        "    d = {'a': 'apple', 'b': 'berry', 'c': 'cherry’}\n",
        "    for key in d:\n",
        "        print(key +\" \"+ d[key])\n",
        "The above would print out:\n",
        "    \n",
        "    a apple\n",
        "    \n",
        "    b berry\n",
        "    \n",
        "    c cherry"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "5d489f26-2434-4ea1-bd14-9014b8233675",
      "metadata": {
        "id": "5d489f26-2434-4ea1-bd14-9014b8233675",
        "outputId": "5a302deb-4029-481d-b9f4-a6b5a51db815"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "3\n",
            "{'EcoRI': 'GAATTC', 'AvaII': 'GG(A|T)CC', 'BisI': 'GC[ATGC]GC'}\n",
            "EcoRI\n",
            "AvaII\n",
            "BisI\n",
            "{'EcoRI': 'GAATTC', 'BisI': 'GC[ATGC]GC'}\n",
            "{'EcoRI': 'GAATTC', 'BisI': 'GC[ATGC]GC', 'PamII': 'GRCG(C|T)C'}\n"
          ]
        }
      ],
      "source": [
        "#create an empty dictionary that associates restriction enzymes with their particular cut sequence\n",
        "# ----------------------------\n",
        "# remember that dictionaries use {} whereas lists use []\n",
        "\n",
        "enzymes ={}\n",
        "\n",
        "#assign key value pairs\n",
        "# we could also have done this in the following way:\n",
        "#enzymes={\"EcoRI\":r\"GAATTC\",\"AvaII\":r\"GG(A|T)CC\",\"BisI\":r\"GC[ATGC]GC\"}\n",
        "\n",
        "enzymes[\"EcoRI\"]=\"GAATTC\"\n",
        "# sneaking in a tiny bit of REGEX - Module5B-  for the next two lines:\n",
        "enzymes[\"AvaII\"]=\"GG(A|T)CC\"\n",
        "enzymes[\"BisI\"]=\"GC[ATGC]GC\"\n",
        "#ask for the length of the dictionary\n",
        "print(len(enzymes))\n",
        "#print out all the key value pairs\n",
        "print(enzymes)\n",
        "for key in enzymes:\n",
        "    print(key)\n",
        "\n",
        "# Remove the \"AvaII\" entry.\n",
        "del(enzymes[\"AvaII\"])\n",
        "print(enzymes)\n",
        "#let's add another restriction enzyme to the dictionary\n",
        "enzymes[\"PamII\"]=\"GRCG(C|T)C\"\n",
        "\n",
        "print(enzymes)"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "aa79cba2-b9c2-4746-9729-60bb3576b459",
      "metadata": {
        "id": "aa79cba2-b9c2-4746-9729-60bb3576b459"
      },
      "source": [
        "# Back to our trinucleotide example..."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "362e448a-ada4-4858-9512-b58265230d04",
      "metadata": {
        "id": "362e448a-ada4-4858-9512-b58265230d04",
        "outputId": "05efb0a4-37f5-4b31-b090-d553320441ab"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "-----------------------\n",
            "{'AAA': 0, 'AAC': 0, 'AAG': 0, 'AAT': 1, 'ACA': 0, 'ACC': 0, 'ACG': 1, 'ACT': 0, 'AGA': 0, 'AGC': 0, 'AGG': 0, 'AGT': 0, 'ATA': 0, 'ATC': 2, 'ATG': 1, 'ATT': 0, 'CAA': 0, 'CAC': 0, 'CAG': 0, 'CAT': 0, 'CCA': 0, 'CCC': 0, 'CCG': 0, 'CCT': 0, 'CGA': 1, 'CGC': 1, 'CGG': 0, 'CGT': 1, 'CTA': 0, 'CTC': 0, 'CTG': 1, 'CTT': 0, 'GAA': 0, 'GAC': 0, 'GAG': 0, 'GAT': 2, 'GCA': 0, 'GCC': 0, 'GCG': 0, 'GCT': 1, 'GGA': 0, 'GGC': 0, 'GGG': 0, 'GGT': 0, 'GTA': 1, 'GTC': 0, 'GTG': 0, 'GTT': 0, 'TAA': 0, 'TAC': 1, 'TAG': 0, 'TAT': 0, 'TCA': 0, 'TCC': 0, 'TCG': 2, 'TCT': 0, 'TGA': 2, 'TGC': 0, 'TGG': 0, 'TGT': 0, 'TTA': 0, 'TTC': 0, 'TTG': 0, 'TTT': 0}\n",
            "-----------------------\n",
            "2\n",
            "CPU times: user 152 µs, sys: 30 µs, total: 182 µs\n",
            "Wall time: 183 µs\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "# same as above but USING A DICTIONARY NOW! --> waaaaay more efficient way to handle the problem\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "# setting up an empty dictionary to fill via the for loop\n",
        "counts={}\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "# you can use for to loop through a dictionary (like we did with lists)\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            counts[trinucleotide]=count\n",
        "print(\"-----------------------\")\n",
        "#Print out counts dictionary\n",
        "print(counts)\n",
        "print(\"-----------------------\")\n",
        "#print counts for a specific trinucleotide\n",
        "print(counts[\"TGA\"])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "0b94f3ab-b5a8-4538-b0eb-42c53f686864",
      "metadata": {
        "id": "0b94f3ab-b5a8-4538-b0eb-42c53f686864",
        "outputId": "d24278ef-05ed-4846-e6dd-b5249115f6d5"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "*************************\n",
            "{'AAT': 1, 'ACG': 1, 'ATC': 2, 'ATG': 1, 'CGA': 1, 'CGC': 1, 'CGT': 1, 'CTG': 1, 'GAT': 2, 'GCT': 1, 'GTA': 1, 'TAC': 1, 'TCG': 2, 'TGA': 2}\n",
            "--------------------\n",
            "2\n"
          ]
        },
        {
          "ename": "KeyError",
          "evalue": "'AAA'",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "File \u001b[0;32m<timed exec>:19\u001b[0m\n",
            "\u001b[0;31mKeyError\u001b[0m: 'AAA'"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "# This code more efficient than above since you get rid of a lot of unnecessary values of the list and only keep the non-0 values\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "counts={}\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            #add in an if statement so that we can get rid of all the trinucleotides have a count of 0\n",
        "            if count>0:\n",
        "                counts[trinucleotide]=count\n",
        "print(\"*\"*25)\n",
        "print(counts)\n",
        "print(\"-\"*20)\n",
        "#print counts for a specific trinucleotide\n",
        "print(counts[\"TGA\"])\n",
        "# What happens now?\n",
        "print(counts[\"AAA\"])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "07d762f2-bad8-47b1-92b4-e7f8d9efd845",
      "metadata": {
        "id": "07d762f2-bad8-47b1-92b4-e7f8d9efd845"
      },
      "source": [
        "Hmmmm. Well, well, well -  what happened here?\n",
        "\n",
        "We seemed to call the dictionary for a value that isn't contained in the dictionary. The code raised an exception since \"AAA\" wasn't included.\n",
        "\n",
        "We have a way to address that issue, though!"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f761852c-c2b9-4a26-8a95-92b5739136cc",
      "metadata": {
        "id": "f761852c-c2b9-4a26-8a95-92b5739136cc",
        "outputId": "03097239-2f51-447c-9702-133bec6314a6"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "ATC : 2\n",
            "GAT : 2\n",
            "TCG : 2\n",
            "TGA : 2\n",
            "{'AAA': 0, 'AAC': 0, 'AAG': 0, 'AAT': 1, 'ACA': 0, 'ACC': 0, 'ACG': 1, 'ACT': 0, 'AGA': 0, 'AGC': 0, 'AGG': 0, 'AGT': 0, 'ATA': 0, 'ATC': 2, 'ATG': 1, 'ATT': 0, 'CAA': 0, 'CAC': 0, 'CAG': 0, 'CAT': 0, 'CCA': 0, 'CCC': 0, 'CCG': 0, 'CCT': 0, 'CGA': 1, 'CGC': 1, 'CGG': 0, 'CGT': 1, 'CTA': 0, 'CTC': 0, 'CTG': 1, 'CTT': 0, 'GAA': 0, 'GAC': 0, 'GAG': 0, 'GAT': 2, 'GCA': 0, 'GCC': 0, 'GCG': 0, 'GCT': 1, 'GGA': 0, 'GGC': 0, 'GGG': 0, 'GGT': 0, 'GTA': 1, 'GTC': 0, 'GTG': 0, 'GTT': 0, 'TAA': 0, 'TAC': 1, 'TAG': 0, 'TAT': 0, 'TCA': 0, 'TCC': 0, 'TCG': 2, 'TCT': 0, 'TGA': 2, 'TGC': 0, 'TGG': 0, 'TGT': 0, 'TTA': 0, 'TTC': 0, 'TTG': 0, 'TTT': 0}\n",
            "_________\n",
            "2\n",
            "0\n",
            "CPU times: user 737 µs, sys: 160 µs, total: 897 µs\n",
            "Wall time: 881 µs\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "# we can impose other criteria on the dictionary - like only print out the values if there are exactly 2 of them\n",
        "# This will only print out the nucleotide\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "counts={}\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            #add in an if statement because we only care about trinucleotides with 2 counts\n",
        "            count=dna.count(trinucleotide)\n",
        "            counts[trinucleotide]=count\n",
        "            if counts[trinucleotide] >= 2:\n",
        "                print(trinucleotide +\" : \"+str(count))\n",
        "\n",
        "# the underlying dictionary is still present of course and we could print it out, if we wanted to\n",
        "print(counts)\n",
        "# we could also print out the counts for any given key\n",
        "print(\"_________\")\n",
        "#print counts for a specific trinucleotide\n",
        "print(counts[\"TGA\"])\n",
        "# ----------------------------\n",
        "# AND NOW WHAT HAPPENS? COMPARE TO CELL ABOVE\n",
        "# ----------------------------\n",
        "print(counts[\"AAA\"])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "29c2aa65-a9bf-423a-8b7b-4de65fabd28b",
      "metadata": {
        "id": "29c2aa65-a9bf-423a-8b7b-4de65fabd28b"
      },
      "source": [
        "## Dictionary Methods\n",
        "* Items can be removed from a dictionary\n",
        "\n",
        "\t\t\tdel(dict_name[key_name])\n",
        "            \n",
        "* The above command, removes the key and value from the dictionary\n",
        "* You can .pop() a value off of a dictionary which returns the value of the key and simultaneously deletes it (we saw this method with lists, too)\n",
        "\n",
        "            dict_name.pop(“key_1”)\n",
        "            \n",
        "* You can replace the value matched to a key by:\n",
        "            \n",
        "            dict_name[key]=new_value\n",
        "* Built in functions for dictionaries include:\n",
        "            \n",
        "            .keys() -  returns an array of the dictionary’s keys\n",
        "\n",
        "            .values() – returns an array of the dictionary’s values\n",
        "            \n",
        "            .get() – takes two arguments but the second argument is optional and it useful if the dictionary does not contain the key that is being searched for. It may not be immediately obvious to you why the .get method is so useful. However, if you accidentally call a key that doesn’t exist, the program will stop and raise an exception which could de-rail your project. By using .get() you can specify the second argument for what to return if the key doesn’t exist so that your program doesn’t just stop\n",
        "            \n",
        "             .sorted() - We can sort key list before looping to impose a specific order on it that can be repeatable\n",
        "             .items() - Instead of returning a single value or a list of values, items method returns a list of paired values so we can iterate over pairs of data (this is a bit weird…) but we have seen something similar with enumerate function in lists. The more technical definition is that the .items method returns an array of tuples which consists of a key/value pair from the dictionary\n",
        "                         - usually see it as:\n",
        "                             example:\n",
        "                    d = {\"Name\": ”DAP”,\"Age\": 00, “Statistics\": True}\n",
        "                    print(d.items())\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">          \n",
        "\n",
        "    This is a common pattern:\n",
        "        for key in my_dict.keys():\n",
        "            value =my_dict.get(key)\n",
        "            #do something with key and value\n",
        "\n",
        "    So common, in fact, that python has a built in feature to handle it:\n",
        "        for key,value in my_dict.items():\n",
        "            #do something with key and value\n",
        "\n",
        "\n",
        "<div class=\"alert alert-block alert-warning\">\n",
        "\n",
        "    print(dict_name[\"Key1\"]) is the same as  print(dict_name.get(\"Key1\"))\n",
        "\n",
        "    But! With .get(), you can specify what to return if key doesn’t exist! For instance, in the nucleotide example, we happened to know that, with how the loop was structured, the trinucleotide would have to be present 0 times if it wasn't in our final dictionary so we could specify the count of \"0\" if the trinucleotide wasn't present.  \n",
        "\n",
        "    print(dict_name.get(\"Key1\",0))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "d731a89b-2596-4e18-8ee9-f360dc13e62f",
      "metadata": {
        "id": "d731a89b-2596-4e18-8ee9-f360dc13e62f",
        "outputId": "dd81e32f-8ffd-4566-e77e-125d64369a76"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "{'house': 'Haus', 'cat': 'Katze', 'red': 'rouge', 'blau': 'bleu'}\n",
            "~~~~~~~~~\n",
            "dict_keys(['house', 'cat', 'red', 'blau'])\n",
            "++++++++++++\n",
            "dict_values(['Haus', 'Katze', 'rouge', 'bleu'])\n",
            "~~~~~~~~~~~~\n",
            "house Haus\n",
            "cat Katze\n",
            "red rouge\n",
            "blau bleu\n",
            "We haven't discussed yellow\n",
            "Katze\n"
          ]
        },
        {
          "ename": "KeyError",
          "evalue": "'yellow'",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[12], line 19\u001b[0m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28mprint\u001b[39m(w\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124myellow\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWe haven\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt discussed yellow\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[1;32m     18\u001b[0m \u001b[38;5;28mprint\u001b[39m(w\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mcat\u001b[39m\u001b[38;5;124m\"\u001b[39m,\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mWe haven\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mt discussed cats\u001b[39m\u001b[38;5;124m\"\u001b[39m))\n\u001b[0;32m---> 19\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mw\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43myellow\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m)\n\u001b[1;32m     20\u001b[0m \u001b[38;5;66;03m# if I used the other method of calling keys, I would raise an exception instead of telling me something useful \u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;66;03m# that geld doesn't isn't a member of this list\u001b[39;00m\n",
            "\u001b[0;31mKeyError\u001b[0m: 'yellow'"
          ]
        }
      ],
      "source": [
        "# A note on the update() method of dictionaries: There are two red:rot key-value pairs and when updating, python removes one of the them.\n",
        "w={\"house\":\"Haus\",\"cat\":\"Katze\",\"red\":\"rot\"}\n",
        "w1 = {\"red\":\"rouge\",\"blau\":\"bleu\"}\n",
        "w.update(w1)\n",
        "print(w)\n",
        "print(\"~~~~~~~~~\")\n",
        "#{'house': 'Haus', 'blau': 'bleu', 'red': 'rouge', 'cat': 'Katze'}\n",
        "print(w.keys())\n",
        "print(\"++++++++++++\")\n",
        "print(w.values())\n",
        "print(\"~~~~~~~~~~~~\")\n",
        "#You can iterate over a dictionary\n",
        "for key in w:\n",
        "    print(key, w[key])\n",
        "\n",
        "# I want to return the fact that he is not a member of this list\n",
        "print(w.get(\"yellow\",\"We haven't discussed yellow\"))\n",
        "print(w.get(\"cat\",\"We haven't discussed cats\"))\n",
        "#print(w[\"yellow\"])\n",
        "# if I used the other method of calling keys, I would raise an exception instead of telling me something useful\n",
        "# that geld doesn't isn't a member of this list"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "e8c2ea95-2339-4262-b97e-fc3d4175b4bd",
      "metadata": {
        "id": "e8c2ea95-2339-4262-b97e-fc3d4175b4bd",
        "outputId": "3637ea0c-ff21-4d47-beee-e47403c3990d"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "{'AAT': 1, 'ACG': 1, 'ATC': 2, 'ATG': 1, 'CGA': 1, 'CGC': 1, 'CGT': 1, 'CTG': 1, 'GAT': 2, 'GCT': 1, 'GTA': 1, 'TAC': 1, 'TCG': 2, 'TGA': 2}\n",
            "Is AAA present? \n",
            "0\n",
            "Here is the sorted function with criteria of 2\n",
            "**********************\n",
            "ATC\n",
            "GAT\n",
            "TCG\n",
            "TGA\n",
            "Here are all the trinucleotides that are present in this string\n",
            "The trinucleotide: AAT is represented 1 times\n",
            "The trinucleotide: ACG is represented 1 times\n",
            "The trinucleotide: ATC is represented 2 times\n",
            "The trinucleotide: ATG is represented 1 times\n",
            "The trinucleotide: CGA is represented 1 times\n",
            "The trinucleotide: CGC is represented 1 times\n",
            "The trinucleotide: CGT is represented 1 times\n",
            "The trinucleotide: CTG is represented 1 times\n",
            "The trinucleotide: GAT is represented 2 times\n",
            "The trinucleotide: GCT is represented 1 times\n",
            "The trinucleotide: GTA is represented 1 times\n",
            "The trinucleotide: TAC is represented 1 times\n",
            "The trinucleotide: TCG is represented 2 times\n",
            "The trinucleotide: TGA is represented 2 times\n",
            "CPU times: user 218 µs, sys: 29 µs, total: 247 µs\n",
            "Wall time: 232 µs\n"
          ]
        }
      ],
      "source": [
        "%%time\n",
        "# Here are some more important methods for dictionaries!\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "#dna=\"AATGATCGA\"\n",
        "counts={}\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            #add in an if statement so that we can get rid of all the trinucleotides\n",
        "            #have a count of 0\n",
        "            if count >0:\n",
        "                counts[trinucleotide]=count\n",
        "print(counts)\n",
        "print(\"Is AAA present? \")\n",
        "# ******************************\n",
        "# INCREDIBLY IMPORTANT METHOD ALERT! .get() allows you to search a dictionary\n",
        "# but also specify what to return in the case that what you are searching\n",
        "# for doesn't exist in the dictionary! This will serve two purposes:\n",
        "# 1. Your program doesn't crash when it doesn't find something that you specified\n",
        "# 2. it gives you information that the item doesn't exist\n",
        "# ******************************\n",
        "print(counts.get(\"AAA\",0))\n",
        "# -------------------------------------\n",
        "# Alert!!! Here is the sorted function in use!\n",
        "# -------------------------------------\n",
        "print(\"Here is the sorted function with criteria of 2\")\n",
        "print(\"**********************\")\n",
        "#for trinucleotide,count in sorted(counts.items()):\n",
        "for trinucleotide,count in counts.items():\n",
        "    if count == 2:\n",
        "        print(trinucleotide)\n",
        "print(\"Here are all the trinucleotides that are present in this string\")\n",
        "for trinucleotide,count in counts.items():\n",
        "    print(\"The trinucleotide: \"+trinucleotide+\" is represented \"+str(count)+\" times\")"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "29e27ad2-0ed8-4675-9195-154799085820",
      "metadata": {
        "id": "29e27ad2-0ed8-4675-9195-154799085820"
      },
      "outputs": [],
      "source": [
        "%%time\n",
        "# Here is an example where I try to incorporate all of the ideas that we have seen so far in this module:\n",
        "dna=\"AATGATCGATCGTACGCTGA\"\n",
        "counts={}\n",
        "all_bases=[\"A\",\"C\",\"G\",\"T\"]\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            count=dna.count(trinucleotide)\n",
        "            #add in an if statement so that we can get rid of all the trinucleotides have a count of 0\n",
        "            if count >0:\n",
        "                counts[trinucleotide]=count\n",
        "\n",
        "print(counts)\n",
        "#print counts for a specific trinucleotide\n",
        "#print(counts[\"TGA\"])\n",
        "for base1 in all_bases:\n",
        "    for base2 in all_bases:\n",
        "        for base3 in all_bases:\n",
        "            trinucleotide=base1+base2+base3\n",
        "            #add in an if statement because we only care about trinucleotides with 2 counts\n",
        "            #count=dna.count(trinucleotide)\n",
        "            #counts[trinucleotide]=count\n",
        "            if counts.get(trinucleotide,0) == 1 or counts.get(trinucleotide,0) == 2:\n",
        "                print(trinucleotide)\n",
        "# Here's a method that we haven't seen yet.\n",
        "print(counts.keys())\n",
        "#print counts for a specific trinucleotide\n",
        "print(counts[\"TGA\"])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "992758d7-931a-4811-bee3-a2e3472fcbb2",
      "metadata": {
        "id": "992758d7-931a-4811-bee3-a2e3472fcbb2"
      },
      "source": [
        "## Lists can be entries - values, not keys - in a dictionary and vice versa\n",
        "* Lists can be an entry in a dictionary\n",
        "\n",
        "* They can't be keys - why not?\n",
        "\n",
        "* When you access a list in a dictionary by:\n",
        "\n",
        "        dict_name[‘list_key’]\n",
        "\n",
        "* This can be a bit challenging and if you want to print out all the information you will need to use a for loop"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "22100612-7c83-4f97-9e9a-9ad4eb5d35d9",
      "metadata": {
        "id": "22100612-7c83-4f97-9e9a-9ad4eb5d35d9",
        "outputId": "4db506f1-1e38-4148-e257-7e81bb9b0bcc"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "p53\n",
            "['cytoplasm', 'nucleus', 'extranuclear']\n",
            "cytoplasm\n",
            "*******\n",
            "p53\n",
            "9606\n",
            "['cytoplasm', 'extranuclear', 'nucleus']\n",
            "mus musculus\n",
            "96020344\n"
          ]
        }
      ],
      "source": [
        "# A dictionary can contain lists:\n",
        "gene={'name':\"p53\", 'taxonomy':\"9606\",\"locations\":[\"cytoplasm\",\"nucleus\",\"extranuclear\"]}\n",
        "print(gene['name'])\n",
        "print(gene['locations'])\n",
        "print(gene['locations'][0])\n",
        "#add keys and values\n",
        "gene['species']='homo sapiens'\n",
        "gene['id']=96020344\n",
        "# we can use mehtods on dictionaries. Let's try sort\n",
        "gene[\"locations\"].sort()\n",
        "# we can replace values in a dictionary:\n",
        "gene['species']='mus musculus'\n",
        "print(\"*******\")\n",
        "for entries in gene:\n",
        "    print(gene[entries])"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "f99f9e40-dee5-4710-88e3-03300d46cb7b",
      "metadata": {
        "id": "f99f9e40-dee5-4710-88e3-03300d46cb7b"
      },
      "outputs": [],
      "source": [
        "#example from codeacademy which has a great section on dictionaries (and other basics of python)\n",
        "#A list of dictionaries\n",
        "lloyd = {\n",
        "    \"name\": \"Lloyd\",\n",
        "    \"homework\": [90.0,97.0,75.0,92.0],\n",
        "    \"quizzes\": [88.0,40.0,94.0],\n",
        "    \"tests\": [75.0,90.0]\n",
        "}\n",
        "alice = {\n",
        "    \"name\": \"Alice\",\n",
        "    \"homework\": [100.0, 92.0, 98.0, 100.0],\n",
        "    \"quizzes\": [82.0, 83.0, 91.0],\n",
        "    \"tests\": [89.0, 97.0]\n",
        "}\n",
        "tyler = {\n",
        "    \"name\": \"Tyler\",\n",
        "    \"homework\": [0.0, 87.0, 75.0, 22.0],\n",
        "    \"quizzes\": [0.0, 75.0, 78.0],\n",
        "    \"tests\": [100.0, 100.0]\n",
        "}\n",
        "#a BIG list of dictionaries\n",
        "students=[lloyd,alice,tyler]\n",
        "#try to print out all the information in the dictionary\n",
        "print(\"-\"*10)\n",
        "print(students)\n",
        "print(\"*\"*10)\n",
        "#we can also print out the information from a specific dictionary\n",
        "print(tyler.items())\n",
        "print(\".\"*10)\n",
        "#this is not really what we want....it's pretty messy\n",
        "for student in students:\n",
        "    print(\"-------\")\n",
        "    print(student[\"name\"])\n",
        "    print(\"-------\")\n",
        "    print(student[\"homework\"])\n",
        "    print(student[\"quizzes\"])\n",
        "    print(student[\"tests\"])"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "ca955d9f-8a70-42e5-a24f-bf2f567f5454",
      "metadata": {
        "id": "ca955d9f-8a70-42e5-a24f-bf2f567f5454"
      },
      "source": [
        "## You can pass a dictionary to a function!"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "83768c86-b549-4680-b4e2-2c11c0f064bd",
      "metadata": {
        "id": "83768c86-b549-4680-b4e2-2c11c0f064bd"
      },
      "outputs": [],
      "source": [
        "# you can also pass dictionaries to functions\n",
        "def average(numbers):\n",
        "    total = sum(numbers)\n",
        "    total = float(total)\n",
        "    print(\"********** HERE I AM IN THE FIRST FUNCTION!!******\")\n",
        "    total=total/len(numbers)\n",
        "    return total\n",
        "\n",
        "def get_average(student):\n",
        "    homework=average(student[\"homework\"])\n",
        "    print(\"***Here I am in the second function!\")\n",
        "    quizzes=average(student[\"quizzes\"])\n",
        "    tests=average(student[\"tests\"])\n",
        "    return 0.1*homework+0.3*quizzes+0.6*tests\n",
        "\n",
        "def get_letter_grade(score):\n",
        "    print(\"Here I am in the third function!\")\n",
        "    if score >= 90:\n",
        "        return \"A\"\n",
        "    elif score >=80:\n",
        "        return \"B\"\n",
        "    elif score >=70:\n",
        "        return \"C\"\n",
        "    elif score >=60:\n",
        "        return \"D\"\n",
        "    else:\n",
        "        return \"F\"\n",
        "\n",
        "# Here is the dictionary that you will then pass in the function\n",
        "lloyd = {\n",
        "    \"name\": \"Lloyd\",\n",
        "    \"homework\": [90.0,97.0,75.0,92.0],\n",
        "    \"quizzes\": [88.0,40.0,94.0],\n",
        "    \"tests\": [75.0,90.0]\n",
        "}\n",
        "#Passing a dictionary, lloyd, to the get_average function which is then passed to the get_letter_grade function\n",
        "print(get_letter_grade(get_average(lloyd)))\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "70b14ad3-ece1-4d65-86e2-8db9da822d16",
      "metadata": {
        "id": "70b14ad3-ece1-4d65-86e2-8db9da822d16",
        "outputId": "3a7646c9-4ab1-49f0-cc81-626ca0faa448"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "0\n",
            "reagent1\n",
            "**********\n",
            "price: 10\n",
            "stock: 6\n",
            "60\n",
            "60\n",
            "reagent2\n",
            "**********\n",
            "price: 25\n",
            "stock: 0\n",
            "0\n",
            "60\n",
            "reagent3\n",
            "**********\n",
            "price: 1\n",
            "stock: 32\n",
            "32\n",
            "92\n",
            "reagent4\n",
            "**********\n",
            "price: 5\n",
            "stock: 10\n",
            "50\n",
            "*\n",
            "-\n",
            "40\n"
          ]
        }
      ],
      "source": [
        "# I have written this function in TWO ways - with one way hashed out\n",
        "# I have two dictionaries: prices and stock\n",
        "# second way, I have one combined dictionary which has the value of a list at each key\n",
        "#def compute_lab_cost(reagent):\n",
        "#    total_here = 0\n",
        "#    for item in reagent:\n",
        "#        #you have access to the elements of a dictionary even though it hasn't been passed into the function\n",
        "#        total_here +=prices[item]\n",
        "#    return total_here\n",
        "def compute_lab_cost(reagent):\n",
        "    total_here = 0\n",
        "    for item in reagent:\n",
        "        total_here +=price_stock[item][0]\n",
        "    return total_here\n",
        "# two dictionaries here: one that tells you the price of each item\n",
        "#prices = { \"reagent1\":10,  \"reagent2\":25, \"reagent3\":1, \"reagent4\":5}\n",
        "# second dictionary that tells you how many of each item you have in stock\n",
        "stock = {\"reagent1\":6,\"reagent2\": 0, \"reagent3\":32, \"reagent4\":10}\n",
        "\n",
        "# We can combine these two dictionaries into one dictionary that has a list for each reagent key\n",
        "# the first element of that list is price and the second element is stock. I have hashed it out below\n",
        "#price_stock={\"reagent1\":[10,6],\"reagent2\":[25,0],\"reagent3\":[1,32],\"reagent4\":[5,10]}\n",
        "\n",
        "\n",
        "total=0\n",
        "for key in prices:\n",
        "    print(total)\n",
        "    print(key)\n",
        "    print(\"**********\")\n",
        "    print(\"price: \"+ str(prices[key]))\n",
        "    print(\"stock: \"+ str(stock[key]))\n",
        "    print(prices[key]*stock[key])\n",
        "    total = total + prices[key]*stock[key]\n",
        "\n",
        "print(\"*\")\n",
        "#print(total)\n",
        "print(\"-\")\n",
        "\n",
        "#list of keys\n",
        "shopping_for_lab = [\"reagent4\", \"reagent2\", \"reagent1\"]\n",
        "\n",
        "# PASSING A DICTIONARY TO A FUNCTION!\n",
        "print(compute_lab_cost(shopping_for_lab))"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "3c19df2a-bc59-4b37-b2d2-9e3ba7e0e3a5",
      "metadata": {
        "id": "3c19df2a-bc59-4b37-b2d2-9e3ba7e0e3a5"
      },
      "source": [
        "# In class question (30 min):\n",
        "\n",
        "Write a program that translates a user-specified nucleotide sequence (in a text file) into amino acids. You will need to use a dictionary to effectively solve this challenge. A textfile with a dictionary has been uploaded with your assignment (called, oddly enough, Assignment_3_dictionary.txt) – you can cut and paste it directly into your program. You can print the resulting amino acid to the screen or to an external file.\n",
        "\n",
        "Your program should use the standard genetic code. Additionally, your program should include checks to ensure that the user has uploaded an appropriate text file. So far, we have seen built-in methods to address accidental user input of lowercase letters. Your solution to problem 3 should additionally be able to elegantly solve the following user input issues: what happens if the file has a number of nucleotides that isn't divisible by 3? What if the file uses non-standard letters instead of the standard genetic code?\n",
        "\n",
        "Besides any other issues that you may dream up for violating your expected input (trying to 'break your code' is a great way to test for robustness!), your code should at least pass the following assertions:\n",
        "assert translate_dna(\"ATGTTCGGT\") == \"MFG\"\n",
        "assert translate_dna(\"ATCGATCGAT\")  == \"IDR\"\n",
        "assert translate_dna(\"ACGANCGAT\") == You can one of the many online translation tools such as the following: http://web.expasy.org/translate/ to confirm your answer for any sequence that you \tmay choose to provide.\n",
        "\n",
        "Hint:  To solve this problem easily, you need to first recognize that there are three \tmajor components/sub-problems that you have to solve.\n",
        "\n",
        "1.\tYou need to split the nucleotide sequence into codons.     \n",
        "-\twe have already learned how to ‘pull out’ parts of strings - now we just need to ‘automate’  proce    ss\n",
        "-\trecognize that each codon is three nucleotides long\n",
        "2.\tTranslate each codon into the appro e amino     acid\n",
        "-\tThe given dictionary would come in handy here\n",
        "3.\tCreate a protein by adding all the amino acids together\n",
        "________\n",
        "gencode = {\n",
        "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
        "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
        "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
        "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
        "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
        "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
        "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
        "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
        "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
        "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
        "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
        "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
        "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
        "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
        "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
        "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "084379a6-f147-4eac-819c-31cb064eabfe",
      "metadata": {
        "id": "084379a6-f147-4eac-819c-31cb064eabfe"
      },
      "outputs": [],
      "source": [
        "my_sequence = input(\"please enter gene sequence: \\n\")\n",
        "#dictionary of codon translations\n",
        "gencode = {\n",
        "    'ATA':'I', 'ATC':'I', 'ATT':'I', 'ATG':'M',\n",
        "    'ACA':'T', 'ACC':'T', 'ACG':'T', 'ACT':'T',\n",
        "    'AAC':'N', 'AAT':'N', 'AAA':'K', 'AAG':'K',\n",
        "    'AGC':'S', 'AGT':'S', 'AGA':'R', 'AGG':'R',\n",
        "    'CTA':'L', 'CTC':'L', 'CTG':'L', 'CTT':'L',\n",
        "    'CCA':'P', 'CCC':'P', 'CCG':'P', 'CCT':'P',\n",
        "    'CAC':'H', 'CAT':'H', 'CAA':'Q', 'CAG':'Q',\n",
        "    'CGA':'R', 'CGC':'R', 'CGG':'R', 'CGT':'R',\n",
        "    'GTA':'V', 'GTC':'V', 'GTG':'V', 'GTT':'V',\n",
        "    'GCA':'A', 'GCC':'A', 'GCG':'A', 'GCT':'A',\n",
        "    'GAC':'D', 'GAT':'D', 'GAA':'E', 'GAG':'E',\n",
        "    'GGA':'G', 'GGC':'G', 'GGG':'G', 'GGT':'G',\n",
        "    'TCA':'S', 'TCC':'S', 'TCG':'S', 'TCT':'S',\n",
        "    'TTC':'F', 'TTT':'F', 'TTA':'L', 'TTG':'L',\n",
        "    'TAC':'Y', 'TAT':'Y', 'TAA':'_', 'TAG':'_',\n",
        "    'TGC':'C', 'TGT':'C', 'TGA':'_', 'TGG':'W'}\n"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3 (ipykernel)",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.11.5"
    },
    "colab": {
      "provenance": []
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}